Lib/test/test_logging.py

```diff
@@ -4321,22 +4321,55 @@ def __del__(self):
                     except Exception:
                         logging.exception("exception in __del__")
 
-            a = A()"""
+            a = A()
+        """)
         rc, out, err = assert_python_ok("-c", code)
         err = err.decode()
         self.assertIn("exception in __del__", err)
         self.assertIn("ValueError: some error", err)
 
+    def test_logging_at_shutdown_open(self):
+        # bpo-26789: FileHandler keeps a reference to the builtin open()
+        # function to be able to open or reopen the file during Python
+        # finalization.
+        filename = os_helper.TESTFN
+        self.addCleanup(os_helper.unlink, filename)
+
+        code = textwrap.dedent(f"""
+            import builtins
+            import logging
+
+            class A:
+                def __del__(self):
+                    logging.error("log in __del__")
+
+            # basicConfig() opens the file, but logging.shutdown() closes
+            # it at Python exit. When A.__del__() is called,
+            # FileHandler._open() must be called again to re-open the file.
+            logging.basicConfig(filename={filename!r})
+
+            a = A()
+
+            # Simulate the Python finalization which sets
+            # modules attributess to None
+            builtins.open = None
```

Comment:
`open` is not set to None. It is removed.

Also some names are removed from the sys module: `path`, `path_hooks`, etc.

Reply:
> open is not set to None. It is removed.

Oh, you're right! The finalize_modules() function (*) is a complex beast. It sets module attributes to None, but sys and builtins are treated differently. The operation which "removes" builtin symbols happens in finalize_restore_builtins():

```
PyDict_Update(interp->builtins, interp->builtins_copy)
```

A copy of the builtins dict is created before "open" is added to the builtins dict, on purpose.

(*) I recently renamed finalize_modules(), it was previously called _PyImport_Cleanup even if it's unrelated to import!